### PR TITLE
feat(linter): oxc/ifs-same-conditions

### PR DESCRIPTION
**Pull request by camc314**
Inspired by https://rust-lang.github.io/rust-clippy/master/#/ifs_same_cond

### CODE REVIEW

**Comment by camc314** [comment]
* **#6150** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6150?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camc314 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/oxc_linter/src/rules/oxc/ifs_same_conditions.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            return;
+        };
+
+        let tests = collect_if_tests(if_stmt);
+        if tests.len() < 2 {
+            return;
+        }
+
+        tests.iter().tuple_combinations().for_each(|(first_test, second_test)| {
+            if first_test.content_eq(second_test) {
```
Comment (camc314) [inline crates/oxc_linter/src/rules/oxc/ifs_same_conditions.rs:74]: `content_eq` doesn't consider `(foo as string)` and `(foo)` the same?
this means that we can't catch cases like `if (foo) {} else if (foo as string) {}`
should `content_eq` use https://github.com/oxc-project/oxc/blob/28da77195b88c672f2260dcbf9773a6b8f8c347d/crates/oxc_ast/src/ast_impl/js.rs#L208-L222 and ignore these?

Regardless, doesn't make sense to change in this PR, just something to consider

File: crates/oxc_linter/src/rules/oxc/ifs_same_conditions.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+use oxc_span::cmp::ContentEq;
+use oxc_span::{GetSpan, Span};
+
+fn ifs_same_conditions_diagnostic(span1: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Unreachable if consequent")
+        .with_help(
+            "This if statement test will always be false, as the previous `if` statement covers the same condition.",
+        )
```
Comment (camc314) [inline crates/oxc_linter/src/rules/oxc/ifs_same_conditions.rs:16]: i expanded on clippy's behaviour a decent amount to catch more edge cases.
maybe we rename this rule something like `no-unreachable-else-if` or something like that?
i think these help diagnostic msg could be improved also. but it's late and my brain is stopping working 🙂

**Comment by DonIsaac** [comment]
I've got a funky feeling that there's a unicorn rule that does this. Could you check to see if it's there?

**Comment by Boshen** [comment]
Is this https://eslint.org/docs/latest/rules/no-dupe-else-if ?

**Comment by DonIsaac** [comment]
That's the one

**Comment by camc314** [comment]
Ah sweet didn't realize this rule was already implemented

**Pull request closed without merging**